JetMETCorrections/Type1MET/interface/JetCleanerForType1METT.h

```diff
@@ -0,0 +1,222 @@
+ *
+ * NOTE: class is templated to that it works with reco::PFJets as well as with pat::Jets of PF-type as input
+ *
+ * \authors Matthieu Marionneau ETH
+ *
+ *
+ *
+ */
+
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+
+#include "DataFormats/Common/interface/Ref.h"
+#include "DataFormats/Common/interface/RefToBase.h"
+#include "DataFormats/JetReco/interface/Jet.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidateFwd.h"
+#include "DataFormats/METReco/interface/CorrMETData.h"
+#include "DataFormats/PatCandidates/interface/Jet.h"
+
+#include "CommonTools/Utils/interface/StringCutObjectSelector.h"
+#include "DataFormats/MuonReco/interface/Muon.h"
+
+#include "JetMETCorrections/Type1MET/interface/JetCorrExtractorT.h"
+
+#include <string>
+#include <type_traits>
+
+
+namespace JetCleanerForType1MET_namespace
+{
+  template <typename T, typename Textractor>
+    class InputTypeCheckerT
+  {
+  public:
+
+    void operator()(const T&) const {} // no type-checking needed for reco::PFJet input
+    bool isPatJet(const T&) const {
+      return std::is_base_of<class pat::Jet, T>::value;
+    }
+  };
+
+  template <typename T>
+    class RawJetExtractorT // this template is neccessary to support pat::Jets
+    // (because pat::Jet->p4() returns the JES corrected, not the raw, jet momentum)
+    // But it does not handle the muon removal!!!!! MM
+    {
+    public:
+
+      RawJetExtractorT(){}
+      reco::Candidate::LorentzVector operator()(const T& jet) const
+	{
+	  return jet.p4();
+	}
+    };
+}
+
+
+template <typename T, typename Textractor>
+class JetCleanerForType1METT : public edm::stream::EDProducer<>
+{
+ public:
+
+  explicit JetCleanerForType1METT(const edm::ParameterSet& cfg):
+    moduleLabel_(cfg.getParameter<std::string>("@module_label")),
+    offsetCorrLabel_(""),
+    skipMuonSelection_(nullptr)
+      {
+       token_ = consumes<std::vector<T> >(cfg.getParameter<edm::InputTag>("src"));
+
+       offsetCorrLabel_ = cfg.getParameter<edm::InputTag>("offsetCorrLabel");
+       offsetCorrToken_ = consumes<reco::JetCorrector>(offsetCorrLabel_);
+
+       jetCorrLabel_ = cfg.getParameter<edm::InputTag>("jetCorrLabel"); //for MC
+       jetCorrLabelRes_ = cfg.getParameter<edm::InputTag>("jetCorrLabelRes"); //for data
+       jetCorrToken_ = mayConsume<reco::JetCorrector>(jetCorrLabel_);
+       jetCorrResToken_ = mayConsume<reco::JetCorrector>(jetCorrLabelRes_);
+       
+       jetCorrEtaMax_ = cfg.getUntrackedParameter<double>("jetCorrEtaMax",9.9);
```

Comment:
this parameter value apparently can change the output results.
If so, the parameter has to be tracked.
If the default is needed, define it with fillDescriptions method https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp